### ISSUE TITLE
Customisable Armor Set Setting and armor material groups 

### ISSUE THREAD

**Issue by sirrobzeroone** [minetest-mods/3d_armor#35]
Currently the armor set is set as all registered armor_items must be worn to be a valid set to recieve the set bonus. This is intresting when you start adding extra armor_items.

I worked up a test branch which impliments customisable set using the settings screen, branch is here
https://github.com/sirrobzeroone/3d_armor/tree/armor-set-clean-up

In the process of doing the above I noticed the armor material is gathered from the armor item name using a string match. Which works great until the armor item no longer has the material in the name or the material in the correct location in the name. So I added some new armor groups which start armor_m_"material name" eg armor_m_wood=1 for all wood armor items. Unfortunatly I still had to do an string match to get material name but it is now on a group name which I think is an improvement.

I made the change backwards compatible so if an armor item dosen't have a group included of armor_m_ then it will fallback to calulating set and bonus using the old code (I probably need to harden that old code a little).

Summary of changes
~ Setting to configure parts that make up armor set - default = "head torso legs feet shield"
~ Setting to adjust armor set multiplier default = 1.1
~ New material armor groups to indicate what the material the armor is made of
~ New code in api.lua to calculate armor set and multiplier
~ New code to init.lua to feed set string into table
~ Armor and shield registrations updated to include armor_m_"materials"
~ Updates to Readme.MD

I've done testing and not uncovered any issues but thought it best to do a seperate branch and as the changes are fairly significant. 

Edit: Initial compare of changes (note I stuffed up the MD formating a few times so readme.MD got 3 fixs after this....):
https://github.com/minetest-mods/3d_armor/commit/beb2150f63e793effb30b3c5f80f5b0d3c8a9176

**Comment by sirrobzeroone** [comment]
I've also been playing around with a version on my local machine which removes the legacy code completely and simply assigns armor_m_unknown to the item. This has some advantages and disadvantages

Advantages
~ Cleaner code as the legacy code bits are all removed
~ If only 1 armor set is affected that is registered it will work using the new code and to end user will appear nothing has changed.

Disadvantages
~ If more than 1 armor set is affected then all those armor pieces are considered made of the same material (unknown) which means they can all be mixed and matched to make up a valid set and get a bonus.

The error message would still be printed to debug listing the item name that doesn't have an armor_m included in the group.

Saying all the above I am leaning towards the above as a slightly better version due to the slight code streamlining that occurs, I think it could make downline maintenance easier in the long run.

**Comment by sirrobzeroone** [comment]
I did a little more testing on the above and have now rolled the usage of unknown into the branch. I did leave a copy of the other api with legacy support in there named api_w_legacy.lua. 

If anyone really wants to you can now test out both versions by simply renaming the api file.

I will continue testing the body of the changes are now included in one spot for the moment, if it looks okay to bundle this change into master I will move all the variables etc to the top of the code block....rather than embedded at the top of the new code. 

I built this version on the legacy support version so the changes look more extensive than they are when compared back to existing base api, although I guess it does show the thinning down well between 1st version vs 2nd version:

```
@@ -261,14 +261,6 @@ armor.set_player_armor = function(self, player)
				local value = def.groups["armor_"..attr] or 0
				attributes[attr] = attributes[attr] + value
			end
			local mat = string.match(item, "%:.+_(.+)$")
			if material.name then
				if material.name == mat then
					material.count = material.count + 1
				end
			else
				material.name = mat
			end
		end
	end	
-- New (Dec 2020) for working out armor_set and multiplier	
@@ -279,47 +271,39 @@ armor.set_player_armor = function(self, player)
	local set_bonus_name
	for loc,item in pairs(worn_armor) do
		local item_mat = string.match(minetest.serialize(minetest.registered_items[item].groups),"armor_m_(%a+)\"]")
		if item_mat ~= nil then	
			for k,name in pairs(armor.config.set_elements)do
				if name == loc then
		for k,set_loc in pairs(armor.config.set_elements)do
			if set_loc == loc then
				if item_mat ~= nil then				
					if set_worn[item_mat] == nil then
						set_worn[item_mat] = 0
						set_worn[item_mat] = set_worn[item_mat] + 1
					else
						set_worn[item_mat] = set_worn[item_mat] + 1
					end
				else
					minetest.debug("WARNING:3d_armor - Registered Armor "..item.." dosen't have armor_m_\"material\" assigned")
					if set_worn["unknown"] == nil then
						set_worn["unknown"] = 0
						set_worn["unknown"] = set_worn["unknown"] + 1
					else
						set_worn["unknown"] = set_worn["unknown"] + 1
					end					
				end
			end
		else
			use_legacy_calc = 1
			minetest.debug("WARNING:3d_armor - Registered Armor "..item.." dosen't have armor_m_\"material\" assigned")
		end

		end		
	end	
		if use_legacy_calc ~= 1 then
			for mat_name,arm_piece_num in pairs(set_worn) do
				if arm_piece_num == #armor.config.set_elements then 
					armor_multi = armor.config.set_multiplier
					set_bonus_name = mat_name
				end
			end	
	for mat_name,arm_piece_num in pairs(set_worn) do
		if arm_piece_num == #armor.config.set_elements then 
			armor_multi = armor.config.set_multiplier
			set_bonus_name = mat_name
		end
	end	
-- End New (Dec 2020)
	for group, level in pairs(levels) do
		if use_legacy_calc ~= 1 then
			if level > 0 then
				level = level * armor.config.level_multiplier			
					if armor_multi ~= 0 then
						level = level * armor.config.set_multiplier
					end
			end
		else
		-- Used for legacy support were armor_item dosen't have material group armor_m_xxx assigned
			if level > 0 then
				level = level * armor.config.level_multiplier			
					if material.name and material.count == #self.elements then
						level = level * armor.config.set_multiplier
					end
		if level > 0 then
			level = level * armor.config.level_multiplier			
			if armor_multi ~= 0 then
				level = level * armor.config.set_multiplier
			end
		end
		local base = self.registered_groups[group]
```

**Comment by sirrobzeroone** [comment]
I'm in two minds over the material groups thing. I see the value in the set customisating setting and set bonus multiplier but not so much on the material groups...well other than consistancy with how other things track material names in minetest.

It may simply be better to well document the current behaviour that material name must be at the end of your item name no mater how long eg randommod:random_name_thats_really_long_for_armor_item_wood

anyways I put the question up the forum board we'll see if any strong direction is given.

https://forum.minetest.net/viewtopic.php?f=11&t=4654&p=386810#p386810

**Comment by sirrobzeroone** [comment]
I've done a branch removing the armor material group and changing the readme to reflect the current behaviour. The code is identical to the other branch except using item_name to derive the material. 

https://github.com/sirrobzeroone/3d_armor/tree/add_armor_set_and_armor_set_bonus_setting

The above is based on the current master that was uploaded 9 hrs ago.

**Comment by sirrobzeroone** [comment]
I'm going to start testing the below branch, I didnt get any feedback via forums, it is however the lower impact change so lower risk. Adding material groups can be done later with basically a 1 line code change and naturally some changes to existing armor registration definitions - but it might be best to maybe do that more formally with clear indication it is a breaking release :) :
https://github.com/sirrobzeroone/3d_armor/tree/add_armor_set_and_armor_set_bonus_setting

I'm going to use mys same test plan as last time as it covers all the basics a bit tedious as I know the code dosent touch these area's but good for regression checking - i thinned it down a little:

- Register armor - P
- Put on and take off armor - P
- Mob damage player - 
  - No armor - 
  - with armor -
- Mob damage armor -
- Player damage player -
  - No armor -
  - with armor -
- Player damage armor -
- Player can't damage armor with incorrect weapon level -
  - Diamond sword on Diamond armor- 
  - Wood sword on Diamond armor - P
- Weapon takes use damage and at correct rate- 
  - Wood sword 10 uses - 
- Craft armor pieces -
  - Diamond Leggings - 
  - Wood Chestplate - 

Mixed armor use testing
- 1 Diamond piece and 1 Wood piece
  - Steel sword on Diamond helmet and Wood chest plate -
  - Mese sword on the above  - 

Specific Change tests
- Adjust set via settings -
- Set bonus recieved for new set settings -
- Set bonus multiplier adjusted via settings -
- Set bonus multiplier applied correctly -
- Warning message printed when item name incorrect -
- "Unknown" set group works as expected -

As before I'll update as I complete testing

**Comment by dacmot** [comment]
This issue looks like it could be closed since PR #41 was merged.
